### DISCUSSION TITLE
RegionAU915AlternateDr() cause "No channel found" error

### DISCUSSION THREAD

**Discussion by QINLIMIN** [Lora-net/LoRaMac-node#1189] [Q&A] [answered]
Hi,
When we set region to AU915, and set channels as "CH_00-07_64". The LoRaMac will prompt "No channel found" error after tried to join and failed at 9th times. After we analysis the code, we found that is due to "RegionAU915AlternateDr()" switch datarate to DR_6 forcibly, DR_6 is not in the datarate range of AU915, which we have debug print the datarate range of AU915 are from DR_0 to DR_5. 
The following snippet is where I print the range in <RegionCommon.c>:
`// debug
printf("datarate range, dr=%d, min=%d, max=%d, i=%d, j=%d \n", 
countNbOfEnabledChannelsParams->Datarate, 
countNbOfEnabledChannelsParams->Channels[i + j].DrRange.Fields.Min, 
countNbOfEnabledChannelsParams->Channels[i + j].DrRange.Fields.Max, i, j);

if( RegionCommonValueInRange( countNbOfEnabledChannelsParams->Datarate,
                                                         countNbOfEnabledChannelsParams->Channels[i + j].DrRange.Fields.Min,
                                                         countNbOfEnabledChannelsParams->Channels[i + j].DrRange.Fields.Max ) == false )
    { // Check if the current channel selection supports the given datarate
        // debug
        printf("datarate out range");
        continue;
    }`
Pls let me understand why switch to DR_6 and the datarate range of AU915 is DR_0 to DR_5? I think it is paradoxical and will cause join failed.
thanks

**Comment by mluis1** [comment, marked as answer]
In order to join a network under Fixed Channel plans such AU915 the specification says that the end-device must make use of 125 KHz and 500 kHz channels.

DR_6 corresponds to a 500 kHz channel.

Therefore, the only possibility to get a `No channel found` notification is if that channel is not enabled within the channel mask.
In your message you say that you enabled channels [0:7]+64 could you check that the channel mask you provide is correct?

**It is not recommended to restrict the number of channels being used.** 

This is the reason why the [Regional Parameters specification](https://lora-alliance.org/wp-content/uploads/2020/11/rp_2-1.0.1.pdf)  specification provides a recommendation on how to select the join channels in an efficient manner.

![image](https://user-images.githubusercontent.com/5763295/137686199-602584d3-2905-4503-a066-70b07bb66593.png)

The AU915 and similar regions are expected to run under 64+8 channels gateways and everything would work smoothly. However, a lot of LoRaWAN networks are only deploying 8 channels gateways. Doing so, introduces drawbacks and everyone must accept the compromises that come with it. The compromise is that one needs to accept that it may take several trials up until we reach the right sub-band. Once the network is joined then everything will work as expected as the Network Server will tell the end-devices what channels it can use.

In case you still would like to restrict the number of channels to 8+1 the right procedure to do so is shown in below code example.
Please note that by applying this restriction you will not be anymore LoRaWAN compliant. Therefore you will not be able to certify your end-devices.

```c
/*!
 * Main application entry point.
 */
int main( void )
{
    BoardInitMcu( );
    BoardInitPeriph( );

...

    if ( LmHandlerInit( &LmHandlerCallbacks, &LmHandlerParams ) != LORAMAC_HANDLER_SUCCESS )
    {
        printf( "LoRaMac wasn't properly initialized\n" );
        // Fatal error, endless loop.
        while ( 1 )
        {
        }
    }

    // Set system maximum tolerated rx error in milliseconds
    LmHandlerSetSystemMaxRxError( 20 );

    // The LoRa-Alliance Compliance protocol package should always be
    // initialized and activated.
    LmHandlerPackageRegister( PACKAGE_ID_COMPLIANCE, &LmhpComplianceParams );

#if defined( REGION_AU915 ) || defined( REGION_US915 )
     // Enabling 1st block of 8 channels (0-7) + channel 64
     uint16_t channelMask[] = { 0x00FF, 0x0000, 0x0000, 0x0000, 0x0001, 0x0000 };
     mibReq.Type = MIB_CHANNELS_MASK;
     mibReq.Param.ChannelsMask = channelMask;
     LoRaMacMibSetRequestConfirm( &mibReq );
     mibReq.Type = MIB_CHANNELS_DEFAULT_MASK;
     mibReq.Param.ChannelsDefaultMask = channelMask;
     LoRaMacMibSetRequestConfirm( &mibReq );
#endif

    LmHandlerJoin( );

    StartTxProcess( LORAMAC_HANDLER_TX_ON_TIMER );

    while( 1 )
    {
...
    }
}
```

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by QINLIMIN** [reply]
Hi Mluis, thanks for your quick reply. 
Now I understand why I got "no channel found" message. 
So in order to comply with AU915 specification, both gateway and end-devices must enable 72 channel masks, `channelMask = {0xffff, 0xffff, 0xffff, 0xffff, 0x00ff, 0x0000}`, which it also can find the channels even it switch DR_6. 
But unfortunately our gateway really only enable 8 channels "CH_00-07_64", that's why we set "channelMask" of end-device to "{ 0x00FF, 0x0000, 0x0000, 0x0000, 0x0000, 0x0000 }". On the other hand, if we enable 64+8 channels on end-devices, they will take more time to find the correct channel as gateway. Anyway we will make end-device comply to certification first.
Pls help to correct me if I'm wrong understanding.

**Reply by mluis1** [reply]
Yes, you understood it well.

However, even a so called 8 channels gateway supports 8+1 channel. All version of SX130x base band chips support this.

Please find below and extract of the settings for AU915 channels used by a SX1302 based gateway.
As you can observe the `chan_Lora_std` modem is setup for the 500 kHz channel while `chan_multiSF_X` handle the 8 125 kHz channels.
```json
        "chan_multiSF_0": {"enable": true, "radio": 0, "if": -400000, "desc": "LoRa channel, 125kHz, all SF, 915.2 MHz"},
        "chan_multiSF_1": {"enable": true, "radio": 0, "if": -200000, "desc": "LoRa channel, 125kHz, all SF, 915.4 MHz"},
        "chan_multiSF_2": {"enable": true, "radio": 0, "if":       0, "desc": "LoRa channel, 125kHz, all SF, 915.6 MHz"},
        "chan_multiSF_3": {"enable": true, "radio": 0, "if":  200000, "desc": "LoRa channel, 125kHz, all SF, 915.8 MHz"},
        "chan_multiSF_4": {"enable": true, "radio": 1, "if": -300000, "desc": "LoRa channel, 125kHz, all SF, 916.0 MHz"},
        "chan_multiSF_5": {"enable": true, "radio": 1, "if": -100000, "desc": "LoRa channel, 125kHz, all SF, 916.2 MHz"},
        "chan_multiSF_6": {"enable": true, "radio": 1, "if":  100000, "desc": "LoRa channel, 125kHz, all SF, 916.4 MHz"},
        "chan_multiSF_7": {"enable": true, "radio": 1, "if":  300000, "desc": "LoRa channel, 125kHz, all SF, 916.6 MHz"},
        "chan_Lora_std":  {"enable": true, "radio": 0, "if":  300000, "bandwidth": 500000, "spread_factor": 8, "implicit_hdr": false, "implicit_payload_length": 17, "implicit_crc_en": false, "implicit_coderate": 1, "desc": "LoRa channel, 500kHz, SF8, 915.9 MHz"},
        "chan_FSK":       {"enable": false,"radio": 0, "if":  300000, "bandwidth": 125000, "datarate":  50000, "desc": "FSK channel, 125kHz, 50kbps, 915.9 MHz"}
```

In this situation your channel mask must be
`uint16_t channelMask[] = { 0x00FF, 0x0000, 0x0000, 0x0000, 0x0001, 0x0000 };`

I would enforce that there is no need to make changes to the channels mask. I think that we can accept to have some failed attempts.
Once the end-device has joined it will receive the channels to be used by the Network Server.

**Reply by QINLIMIN** [reply]
Yes, thank you very much

**Reply by mluis1** [reply]
If you don't mind I mark the topic as answered.

**Reply by QINLIMIN** [reply]
yes
